pathExists: Return false on "/nix/store" in pure mode

edolstra opened the pull request:
# Motivation

`AllowListInputAccessor` has the invariant that if a path is accessible, its parent directories are also considered accessible (though reading them only yields the allowed subdirectories). As a result `builtins.pathExists "/nix/store"` returns true.

However this wasn't the behaviour of previous path access control, where `builtins.pathExists "/nix/store"` returns false even if a subdirectory of the store is accessible.

Fixes #9672.

<!-- Briefly explain what the change is about and why it is desirable. -->

# Context
<!-- Provide context. Reference open issues if available. -->

<!-- Non-trivial change: Briefly outline the implementation strategy. -->

<!-- Invasive change: Discuss alternative designs or approaches you considered. -->

<!-- Large change: Provide instructions to reviewers how to read the diff. -->

# Priorities and Process

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

Ericson2314 commented:
What about `builtins.pathExists "/nix"`? Can you test that too? I think we need a more sophisticated check.

Ericson2314 reviewed (changes requested):
What about `builtins.readFileType`?

roberth commented:
> `AllowListInputAccessor` has the invariant

This seems like a good invariant in normal use when composing accessors and such.

> However this wasn't the behaviour of previous path access control,

Maybe the system accessor should be a composition of `AllowListAccessor` and something like `PartialAccessor`?

Regardless of how the accessor is named and structured, this is better handled in the accessor abstraction than in the language frontend, as highlighted by John's observation; this behavior needs to be reflected in all operations. Then the builtins can just use that and not have to duplicate a "hack".

edolstra commented:
I removed the special case handling in `pathExists` and reverted to the old behaviour, i.e. access to a parent is forbidden unless it's explicitly permitted.

roberth commented on src/libfetchers/git-utils.cc:
```diff
@@ -943,17 +943,23 @@ ref<InputAccessor> GitRepoImpl::getAccessor(const Hash & rev, bool exportIgnore)
+
+    /* Grant access to the parent directories of every accessible
+       file. The root is always accessible. */
+    std::unordered_set<CanonPath> files{CanonPath::root};
+    for (auto path : wd.files) {
+        while (!path.isRoot()) {
+            if (!files.insert(path).second) break;
+            path.pop();
+        }
+    }
```
This is a generic solution for getting the closure of path parents. Would be nice to factor out.

roberth commented on src/libfetchers/filtering-input-accessor.hh:
```diff
@@ -64,9 +66,16 @@ struct AllowListInputAccessor : public FilteringInputAccessor
     virtual void allowPrefix(CanonPath prefix) = 0;
 
+    /**
+     * Grant access to a path (but not anything underneath the path,
+     * if it's a directory).
```
What about access to the parents? I think you accidentally wrote the wrong direction here.

The words for path containment are confusing. Maybe that happened? E.g. a path prefix is contained in a more specific path, but a more specific path contains fewer things.
Parent/child terms are unambiguous, but then I don't have a good name for the "parent or same" kind of relation. I must be forgetting something, but using parent/child is always good when they apply.

edolstra commented on src/libutil/source-accessor.hh:
```diff
@@ -111,7 +111,7 @@ struct SourceAccessor
     };
 
-    Stat lstat(const CanonPath & path);
+    virtual Stat lstat(const CanonPath & path);
```
This is necessary to get the correct error message for inaccessible paths, i.e. `forbidden in restricted mode` rather than `does not exist`.

Ericson2314 replied:
I was thinking more the other thing, having `lstatMaybe` not allow `/nix` and `/nix/store` again.

Ericson2314 commented:
We talked about this a while today in the Nix meting. I unfortunately basically changed my mind: I am not comfortable with `/foo/bar` existing *not* implying `/foo` existing, especially while the `SourceAccessor` interface is so new and we are still figuring out what its "laws" are. 

This means I am actually back to preferring the original version of the PR, which did not change the deep abstraction of `SourceAccessor` and just did a quick hack in the evaluator. In particular, I would like to *just* have that quick hack for "restricted eval", and leave pure eval as is, with `builtins.pathExists "/nix/store" == true`

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/2024-09-25-nix-team-meeting-minutes-181-180/52712/1
